cpp/src/decisiontree/batched-levelalgo/builder.cuh

```diff
@@ -0,0 +1,131 @@
+               const DataT* quantiles, IdxT* rowids, IdxT* colids,
+               int n_sampled_rows, int unique_labels,
+               const DecisionTreeParams& params, cudaStream_t stream,
+               std::vector<SparseTreeNode<DataT, LabelT>>& sparsetree,
+               IdxT& num_leaves, IdxT& depth) {
+  Builder<Traits> builder;
+  size_t d_wsize, h_wsize;
+  builder.workspaceSize(d_wsize, h_wsize, params, data, labels, nrows, ncols,
+                        n_sampled_rows, IdxT(params.max_features * ncols),
+                        rowids, colids, unique_labels, quantiles);
+  MLCommon::device_buffer<char> d_buff(d_allocator, stream, d_wsize);
+  MLCommon::host_buffer<char> h_buff(h_allocator, stream, h_wsize);
+  MLCommon::host_buffer<Node<DataT, LabelT, IdxT>> h_nodes(h_allocator, stream,
+                                                           builder.maxNodes);
+  builder.assignWorkspace(d_buff.data(), h_buff.data());
+  builder.train(h_nodes.data(), num_leaves, depth, stream);
+  CUDA_CHECK(cudaStreamSynchronize(stream));
+  d_buff.release(stream);
+  h_buff.release(stream);
+  convertToSparse<Traits>(builder, h_nodes.data(), sparsetree);
+  h_nodes.release(stream);
+}
+
+/**
+ * @defgroup GrowTree Main entry point function for batched-level algo to build
+ *                    a decision tree
+ *
+ * @tparam DataT  data type
+ * @tparam LabelT label type
+ * @tparam IdxT   index type
+ *
+ * @param[in]  d_allocator    device allocator
+ * @param[in]  h_allocator    host allocator
+ * @param[in]  data           input dataset [on device] [col-major]
+ *                            [dim = nrows x ncols]
+ * @param[in]  ncols          number of features in the dataset
+ * @param[in]  nrows          number of rows in the dataset
+ * @param[in]  labels         labels for the input [on device] [len = nrows]
+ * @param[in]  quantiles      histograms/quantiles of the input dataset
+ *                            [on device] [col-major]
+ *                            [dim = params.n_bins x ncols]
+ * @param[in]  rowids         sampled rows [on device] [len = n_sampled_rows]
+ * @param[in]  colids         sampled cols [on device]
+ *                            [len = params.max_features * ncols]
+ * @param[in]  n_sampled_rows number of sub-sampled rows
+ * @param[in]  unique_labels  number of classes (meaningful only for
+ *                            classification)
+ * @param[in]  params         decisiontree learning params
+ * @param[in]  stream         cuda stream
+ * @param[out] sparsetree     output learned tree
+ * @param[out] num_leaves     number of leaves created during tree build
+ * @param[out] depth          max depth of the built tree
+ * @{
+ */
+template <typename DataT, typename LabelT, typename IdxT>
+void grow_tree(std::shared_ptr<MLCommon::deviceAllocator> d_allocator,
+               std::shared_ptr<MLCommon::hostAllocator> h_allocator,
+               const DataT* data, IdxT ncols, IdxT nrows, const LabelT* labels,
```

Comment:
What happens when the user passes in class labels as float32? Would `grow_tree<float, float>` be called? This would be bad, since we are treating a classification problem as a regression problem.

Reply:
Even in the current backend C++ code, the way to distinguish between classification and regression is based on the type of target variable. We can probably add such a check in Python layer.